packages/@eventual/core/src/entity/entity.ts

```diff
@@ -0,0 +1,446 @@
+   *
+   * @param key - key or {@link CompositeKey} of the value to retrieve.
+   */
+  get(key: CompositeKey<Attr, Partition, Sort>): Promise<Attr | undefined>;
+  /**
+   * Get a value and metadata like version.
+   * If your values use composite keys, the namespace must be provided.
+   *
+   * @param key - key or {@link CompositeKey} of the value to retrieve.
+   */
+  getWithMetadata(
+    key: CompositeKey<Attr, Partition, Sort>
+  ): Promise<EntityWithMetadata<Attr> | undefined>;
+  /**
+   * Sets or updates a value within an entity and optionally a namespace.
+   *
+   * Values with namespaces are considered distinct from value without a namespace or within different namespaces.
+   * Values and keys can only be listed within a single namespace.
+   */
+  set(entity: Attr, options?: EntitySetOptions): Promise<{ version: number }>;
+  /**
+   * Deletes a single entry within an entity and namespace.
+   */
+  delete(
+    key: CompositeKey<Attr, Partition, Sort>,
+    options?: EntityConsistencyOptions
+  ): Promise<void>;
+  /**
+   * List entries that match a prefix within an entity and namespace.
+   *
+   * If namespace is not provided, only values which do not use composite keys will be returned.
+   */
+  query(
+    key: QueryKey<Attr, Partition, Sort>,
+    request?: EntityQueryOptions
+  ): Promise<EntityQueryResult<Attr>>;
+  stream(
+    name: string,
+    options: EntityStreamOptions<Attr, Partition, Sort>,
+    handler: EntityStreamHandler<Attr, Partition, Sort>
+  ): EntityStream<Attr, Partition, Sort>;
+  stream(
+    name: string,
+    handler: EntityStreamHandler<Attr, Partition, Sort>
+  ): EntityStream<Attr, Partition, Sort>;
+}
+
+export interface EntityTransactItem<
+  Attr extends Attributes = any,
+  Partition extends CompositeKeyPart<Attr> = CompositeKeyPart<Attr>,
+  Sort extends CompositeKeyPart<Attr> | undefined =
+    | CompositeKeyPart<Attr>
+    | undefined
+> {
+  entity: Entity<Attr, Partition, Sort> | string;
+  operation:
+    | EntitySetOperation<Attr>
+    | EntityDeleteOperation<Attr, Partition, Sort>
+    | EntityConditionalOperation<Attr, Partition, Sort>;
+}
+
+export interface EntitySetOperation<Attr extends Attributes = any> {
+  operation: "set";
+  value: Attr;
+  options?: EntitySetOptions;
+}
+
+export interface EntityDeleteOperation<
+  Attr extends Attributes = any,
+  Partition extends CompositeKeyPart<Attr> = CompositeKeyPart<Attr>,
+  Sort extends CompositeKeyPart<Attr> | undefined =
+    | CompositeKeyPart<Attr>
+    | undefined
+> {
+  operation: "delete";
+  key: CompositeKey<Attr, Partition, Sort>;
+  options?: EntitySetOptions;
+}
+
+/**
+ * Used in transactions, cancels the transaction if the key's version does not match.
+ */
+export interface EntityConditionalOperation<
+  Attr extends Attributes = any,
+  Partition extends CompositeKeyPart<Attr> = CompositeKeyPart<Attr>,
+  Sort extends CompositeKeyPart<Attr> | undefined =
+    | CompositeKeyPart<Attr>
+    | undefined
+> {
```

Comment:
Why do the operation types need default type arguments?

Reply:
Just to avoid having to provide type parameters when there are used later.

![image](https://github.com/functionless/eventual/assets/289213/016f01ef-9fd4-474c-a9a3-956cf40f9274)